hud/src/androidLibMain/kotlin/AndroidHUD.kt

```diff
@@ -164,26 +164,30 @@ class AndroidHUD private constructor(@LayoutRes viewResId: Int, hudConfig: HudCo
 
     private fun subscribeIfNeeded(uiContextData: UiContextObserver.UiContextData?) {
         if (uiContextData != null) {
-            dialogState.observe(uiContextData.lifecycleOwner, Observer {
-                MainScope().launch {
+            MainScope().launch {
+                dialogState.observe(uiContextData.lifecycleOwner, Observer {
                     when (it) {
                         is DialogState.Visible -> loadingDialog.show(uiContextData.fragmentManager, "Kaluga.HUD")
                         is DialogState.Gone -> loadingDialog.dismiss()
                     }
-                }
-            })
+                })
+            }
         }
     }
 
     override val isVisible get() = loadingDialog.isVisible
 
     override fun present(animated: Boolean, completion: () -> Unit): HUD = apply {
-        loadingDialog.presentCompletionBlock = completion
-        dialogState.value = DialogState.Visible
+        MainScope().launch {
+            loadingDialog.presentCompletionBlock = completion
+            dialogState.value = DialogState.Visible
```

Comment:
In case you were wondering `dialogState.postValue(DialogState.Visible)` also works (without the need for `MainScope`), but this is also good.

Reply:
I had a crash using postValue in WTS. My assumption is it happens if we try to show and hide HUD within the same runloop. I'm not that deep in Android but it looks like `set` - sets value instantly, but `post` will perform in the next run loop in the main thread. 
It could be a better solution than changing `post` to `set` but it is quite simple, clear what it does and it just works.

Reply:
But make sense to add tests for this use-case

Reply:
Was the crash in WTS or in Kaluga?

`MainScope().launch` is also asynchronous:

[Launches a new coroutine without blocking the current thread](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/launch.html)

So this is effectively the same as using `postValue`. Launch might be a bit faster in practise (when there is no other coroutines using the main dispatcher), but don't depend on that in your code.

Reply:
The crash was in HUD in Kaluga when we were using it from WTS

Reply:
@thoutbeckers Please check tests I've added. It fails with `post` and passes with `MainScope().launch'

Reply:
Please check my [comments on your test](https://github.com/splendo/kaluga/pull/140#issuecomment-627182545).

Reply:
Also I assume if you change your test to post, it fails but does not crash like in WTS. 

It'd be good to see the crash log of that.

Reply:
No, it crashes. So Probably should catch the exception in tests

Reply:
and no, if I wrap present/dismiss in coroutine scope, it crashes as well

Reply:
#137 - PR which solves WTS crash

Reply:
If the test crashes with `post`, there's probably a bug then in the code elsewhere. Not using post will just hide the bug instead of fixing it. Can you paste the stacktrace?

Reply:
The problem is because `present` doesn't trigger an observer so we try to dismiss dialog which was newer shown. So it looks like a race condition between MainCorution scope and TaskExecutor inside a LiveData.

I think the best solution would be just don't use coroutines to execute code in the main thread (we don't call suspended methods anyway). Another way is to check if the dialog was already presented and only, in this case, dismiss it.

Reply:
Not using coroutines won't magically make the race condition go away.

First of all it's a question of contract. Are dismiss / show asynchronous or synchronous? Are they UI methods that can only be called from the main thread, or can any thread call them?

> Another way is to check if the dialog was already presented and only, in this case, dismiss it.

This indeed seems to be a bug in the implementation in any case, it should not throw some random exception. But fixing the bug means in your case you called dismiss before the dialog was there, so even it will not crash, the dialog will perhaps never be dismissed.

Reply:
@dasdranagon still waiting for the stacktrace, is it similar to #134 ?